Avoid symlink loops when finding files

tomas-abrahamsson opened the pull request:
This addresses #3. This PR adds a `find_files/1` in the helper script, which is like `filelib:wildcard/1`, but differs in that it keeps track of visited files and directories to avoid traversing the same (sub-)directory structure more than once. The `find_files/1` accepts roughly the same patterns as filelib:wilcdard, at least for the patterns that were used. I searched for calls to `filelib:wildcard/1` in the helper and `**` patterns in `.el` files.

tomas-abrahamsson committed:
Avoid symlink loops when finding files

In the helper script, add a find_files/1 function as an alternative to
filelib:wildcard/1.  The find_files/1 keeps track of visited files and
directories to avoid visiting them more than once.

Consider a symlink that point rootwards like this:

    $ ls -l some/sub/dir/
    loop -> ../..
    x.erl

The symlink structure forms a loop and the filelib:wilcard/1 will find
x.erl multiple times, and if the project is large and the symlinks
point many levels back towards the root, filelib:wildcard/1 can take
very long and consume much memory.

Btw, the filelib:wilcard/1 will indeed terminate because the kernel
will eventually return an ELOOP error when the number of followed
symlinks is too high.

tomas-abrahamsson commented:
At the moment, this is a bit of work-in-progress, mainly to get some feedback. I should add some tests, but haven't gotten around to do that yet. I suppose yaemep_SUITE.erl may be a suitable place? So far, I have only tested it manually, also on the large project where I first observed #3. (It took 20 minutes to enumerate all `.erl` and `.hrl` files on a "local" disk in some virtual container, adding `-mode(compile).` only reduced it to 16 minutes, so maybe it is not enough of savings.)

The added main/1 function clause was mainly for manual testing but may turn out useful also in a test suite.

_Edit:_ I should also say that I have not yet used it from emacs and yaemep. I need to try that too.

kjellwinblad commented:
Thank you for the pull request!

> I suppose yaemep_SUITE.erl may be a suitable place?

Yes, that file is a suitable place to add tests for this.

> (It took 20 minutes to enumerate all .erl and .hrl files on a "local" disk in some virtual container, adding -mode(compile). only reduced it to 16 minutes, so maybe it is not enough of savings.)

I think it would be interesting to see how fast `filelib:wildcard` can enumerate `.erl` and `.hrl` files in the same setup but with the difference that symlinks have got removed. That way, we should be able to see if the new `find_files/1` function is significantly slower than `filelib:wildcard`. Is it possible to run this experiment?

> The find_files/1 accepts roughly the same patterns as filelib:wilcdard, at least for the patterns that were used.

The user can configure the search pattern that is used to find files that YAEMEP passes to `etags` by changing the variable  `yaemep-etags-auto-gen-search-pattern` (see the project's README.md file). The user also has the option to change this pattern when clicking the "Project: Generate TAGS" menu item in the "Erlang YAEMEP" menu. Therefore, it would be good if `find_files/1` uses `filelib:wilcdard` as a fallback if it finds a pattern it cannot handle. Can this be done without too much effort?

I will look at the code and make comments inline if I find something else to comment on.

tomas-abrahamsson commented:
> I think it would be interesting to see how fast `filelib:wildcard` can enumerate `.erl` and `.hrl` files in the same setup but with the difference that symlinks have got removed. That way, we should be able to see if the new `find_files/1` function is significantly slower than `filelib:wildcard`. Is it possible to run this experiment?

Yes, no problems, will try to do this.

> The user can configure the search pattern ... Therefore, it would be good if `find_files/1` uses `filelib:wilcdard` as a fallback if it finds a pattern it cannot handle. Can this be done without too much effort?

Ah, I had missed to check for that. Thansk for bringing it to attention. Yes, it should be doable, I will fix.

kjellwinblad commented on emacs_erlang_yaemep_support.erl:
```diff
@@ -92,6 +94,206 @@ erlang_project_dir(ErlangFilePath) ->
+
+is_in_set(Elem, Set) ->
+    sets:is_element(Elem, Set).
+
```
[I have heard people saying](http://erlang.org/pipermail/erlang-questions/2017-April/092195.html) that the maps module is faster than the sets module. The maps module is already used in this file so changing to maps will not limit which Erlang versions YAEMEP can be used with.

tomas-abrahamsson replied:
Maps are indeed faster. Maps or sets was actually one thing I was considering when writing this initially, but forgot to ask about it. Thanks for the answer! Fixed.

kjellwinblad commented on emacs_erlang_yaemep_support.erl:
```diff
@@ -92,6 +94,206 @@ erlang_project_dir(ErlangFilePath) ->
+    Visited0 = new_set(), % containing paths for files, dirs and symlinks
+    Acc0 = [],            % matching files
+    {_Visited, Acc} = traverse_fs(TopDir, Pattern, RT, Visited0, Acc0),
+    lists:reverse(Acc).
```
I think it should be easy to detect if `find_files` is given a pattern that it cannot handle and make it fallback on `filelib:wildcard` in that case (e.g., by making use of a regular expression). That will, of course, not solve the issue with symlink loops when such patterns are given, but I think it is worth the effort since such patterns should be rare and YAEMEP will then still be able to handle such patterns when symlink loops do not exist.

tomas-abrahamsson replied:
Done. (Though not with a regexp... Don't know exactly how such a regexp would look like)

kjellwinblad commented on emacs_erlang_yaemep_support.erl:
```diff
@@ -92,6 +94,206 @@ erlang_project_dir(ErlangFilePath) ->
+    case file:get_cwd() of
+        {ok, Cwd}  -> Cwd;
+        {error, _} -> "."
+    end.
```
Is this function ever called by YAEMEP or is it only intended for testing of `find_files`? I think the path that `file:get_cwd()`will be whatever the current working directory is for emacs.

tomas-abrahamsson replied:
Yes, good point, this makes the code much simpler. Done.

tomas-abrahamsson commented:
I pushed a few fixup comments addressing the comments. I can squash them when things look ok.

With all symlinks pointing within the project (loops or not) (file system cache warmed up):
* `filelib:wildcard` took 7m 3s
* The original version of this PR took 12m 53s (interpreted)
* This PR, but with maps instead of the `sets` module: 7m 33s (interpreted)
* Same but with `-mode(compile).`: 6m 45s

About the 12min 53s vs previously reported 20 minutes: The only difference should basically be that the symlinks were now gone (or possibly some virtual container or physical host load difference). I should ideally verify this, adding back all the symlinks and re-run, but it is kind of slow, and perhaps not that interesting. The main take away is that maps made a lot of difference, thanks for catching this!

_Edit:_ Still to do: automated tests and manual end-to-end tests with emacs.

kjellwinblad commented:
Good, it seems like there is not so much overhead compared to `filelib:wildcard`.

I think I found a bug when I tried to traverse the Erlang/OTP source code with your new function:

```
$ escript emacs_erlang_yaemep_support.erl find_files '/home/user/source/otp/**/*.{erl,hrl}'
Result for "/home/user/source/otp/**/*.{erl,hrl}":
escript: exception error: {function_clause,
                     [{local,str_to_binary,
                          [[47,104,111,109,101,47,107,106,101,108,108,47,115,
                            114,99,47,119,111,114,107,47,111,116,112,47,108,
                            105,98,47,115,115,104,47,116,101,115,116,47,115,
                            115,104,95|...],
                           <<>>]}]}
```

The function seems to work with other Erlang projects so the error is probably caused by some strange file name inside the Erlang/OTP source code directory. Let me know if you can't reproduce the bug.

As for testing, I think you can create a test project with symlink loops inside the test folder.

Do you experience any problems when working with the project that has so long file system traversal times? With project folder traversal times of approximately 7 minutes, my guess is that YAEMEP will more or less constantly traverse the file system as a traversal is triggered every time one saves or opens a ".erl"-file (a new traversal will not start if one is already ongoing). However, this may not be a big problem as the traversal will happen in the background (most of the times*) and YAEMEP should be responsive anyway (at least after the first traversal has completed). I think traversal times of 7 minutes are quite uncommon (traversing the whole Erlang/OTP source code directory just takes a few seconds on my computer).

*[This line](https://github.com/tomas-abrahamsson/yaemep/blob/bad72ddd517599767c528e47df4dbac74601386d/emacs_erlang_yaemep_support.erl#L643) will be problematic with very long traversal times. It is triggered when the user triggers completion of functions in a module that is not in the cache. I think it could be a good idea to have a timeout of 5 seconds or so for this line or all code that should give a list of completions to avoid that Emacs gets blocked.

tomas-abrahamsson committed:
Time out after 5 seconds when finding some files

When looking for files not in the cache, time out after 5 seconds, to
avoid emacs hanging for long, even when project directory traversal
times are very long.

tomas-abrahamsson commented:
I've pushed (and force-pushed) and think I've addressed all issues, but would like to test manually some more.

> I think I found a bug when I tried to traverse the Erlang/OTP source code with your new function ...

Fixed. (<emoji for a bit embarrassed :) >)

> As for testing, I think you can create a test project with symlink loops inside the test folder.

Done. And added a test case.

> Do you experience any problems when working with the project that has so long file system traversal times? ...

Will do some more manual testing to see about this

> [*This](https://github.com/tomas-abrahamsson/yaemep/blob/bad72ddd517599767c528e47df4dbac74601386d/emacs_erlang_yaemep_support.erl#L643) line will be problematic with very long traversal times. ...

Added a [`find_files_with_timeout/2` function](https://github.com/tomas-abrahamsson/yaemep/blob/7201a5cf/emacs_erlang_yaemep_support.erl#L126-L135) in a separate commit.

> I think it could be a good idea to have a timeout of 5 seconds or so for this line or all code that should give a list of completions to avoid that Emacs gets blocked.

Which more places would that be, is it basically all places where find_files is called? And can this happen when just typing in Emacs, or is it triggered by some specific key combination (`M-TAB`)? I was thinking if it can happen while just typing, it may be important to also do it asynchronously in the elisp code. Maybe it already is? (But I also agree at timeout in the erlang support escript code also seems reasonable)

kjellwinblad commented:
```
$ time escript emacs_erlang_yaemep_support.erl find_files '/home/user/source/otp/**/*.{erl,hrl}'
...
real    0m1,559s
user    0m2,371s
sys     0m0,296s
```
Seems to be fast enough.

> Will do some more manual testing to see about this

Looking forward to hearing what you find. You will only get completions for Erlang/OTP modules and functions as well as file local functions until the caches have been updated (should take about 7 minutes for your project). It will also take some time until go-to-thing-at-point starts to work because it will take some time for the TAGS file to get generated.

> Which more places would that be, is it basically all places where find_files is called?

I think the place that you have already fixed was the only crucial place. All other important places that call `find_files` are called asynchronously from elisp code.

> And can this happen when just typing in Emacs, or is it triggered by some specific key combination (M-TAB)?

The `emacs_erlang_yaemep_support.erl` script can be called from Emacs when one is typing if one is using, for example, company-mode and has not turned off automatic completion. This is why it is important that the code that is supposed to a return list of completion options return quite quickly. I think it is up to the completion framework (e.g., company-mode) to decide how the elisp function that generates completion options is executed (I don't think company-mode runs this function in a separate thread or anything like that). As a way to make YAEMEP more robust in case the code that generates completion lists runs for a longer time than expected, one could make the completion list generating code time-out after a certain amount of time but I think this is outside the scope of this pull-request.

tomas-abrahamsson commented:
It was a while ago, a lot of other work got in the way, and I did not spend as much time as I first thought in that huge repo. Anyway, I took it for a brief smoke test today, and things did look ok, nothing broke, and things seems to work. At least I think this change does not seem to make things worse. If it would be ok with you, I could squash the many fixup commits to tidy this PR up a bit?

kjellwinblad commented:
Thanks for the response!

> If it would be ok with you, I could squash the many fixup commits to tidy this PR up a bit?

Yes, that would be great.

I think it would be good to have some basic unit tests in `yaemep_SUITE.erl` for the new `find_files/1` function that will replace `filelib:wildcard` in YAMEP. It would be great if you have the time to add such tests as well. You can add a small test project with symlinks in the test directory if you need that. I'm ready to merge this pull request when we have such tests.

tomas-abrahamsson commented:
Oh, you're absolutely right about unit tests, I added a bunch more of those. And squashed.  (I saved the individual commits at https://github.com/tomas-abrahamsson/yaemep/tree/piece-wise-avoid-symlink-loops in case you would want to see a delta)

kjellwinblad commented:
Thanks for your excellent work! I have merged.

tomas-abrahamsson commented:
Super, thanks!
